README
*********************************************************************************************************
NOTES ON TAF code :
	- basic recipes & doc
  ( releases -> moved to independant file releases.txt )
	- examples
	- comments on features
	- to do list for developers
*********************************************************************************************************
Contact: baudot@in2p3.fr, auguste.besson@iphc.cnrs.fr

Web page: http://www.iphc.cnrs.fr/TAF.html

*********************************************************************************************************
BASIC RECIPES & DOC
*********************************************************************************************************

Before doing anything, set the configuration:
- cd to the taf directory
- source Scripts/thistaf.sh (no configuration needed, except if asked during sourcing)
      (note the old method: source Script/TAF-config is not available anymore)

Documentation is avalibale under the 'doc" directory
- taf_shortDoc.pdf : provides basic help on TAF
- maf_doc.pdf      : provides a detailed description of a former but similar software named MAF
- ClassIndex.html  : provides web-based browsing of classes
  -> You may generate the HTML doc yoursel using the Scripts/makeHTML.doc scripts, read instructions in the script source file.

To run the software:
- simply issue the command: TAF
- first thing to do in TAF:
MimosaAnalysis *gTAF = new MimosaAnalysis() // optional, depends on your system
gTAF->InitSession([runNumber])
gTAF->SetDebug([level])
gTAF->Help();

Note that TAF accepts arguments:
- run TAF --help to know more
- run TAF on a specific run, use
  TAF -run runNumber
- run TAF on a specific run but a generic config file, use
  TAF -run runNumber -cfg ./myConfigDir/myConfigFile.cfg
- run TAF on a specific run and with config file in a specific directory, used
  TAF -run runNumber -cfgdir myConfigDir

To compile the software:
The prefered way is to use the command (from anywhere): maketaf
- Method 1) (equivalent to maketaf script)
> cd code; make
- Method 2) only since v2.6.1 <== deprecated, not checked since ages!!!
> root Scripts/compiltaf.C

To copy the software
- use the command: . Script/tarTAF
- it creates a ttaf_new.tar.gzip with all what you need to export elsewhere


*********************************************************************************************************
EXAMPLES
*********************************************************************************************************

A number of configuration files are provided as examples inside the config_TEST directory. They shall work anywhere,
provided you have the raw data, contact us to get a sample of them. In order to test them you just need to copy the
corresponding file inside your config directory, and change the DataPath variable to the Path where the raw data is
located

- RUN 1040, June 2014
 2 MIMOSA 28, 1 MIMOSA 22THRb, 2 MIMOSA 18, 2 MIMOSA 28 at LNFrascati - Beam Test Facility

- RUN 34539, August 2013
 6 MIMOSA 26 (digital output) with one MIMOSA 34

- RUN 32824, August 2012
 6 MIMOSA 26 (digital output) with one MIMOSA 32 (binary output) and
  the synchronization of two BoardReaders.

- RUN 26000, February 2012
 2 MIMOSA 26 sensors data taken in lab, with no source, and with power pulsing,
 frames with trigger information indicates when the power was switched ON/OFF.

- RUN 226031, November 2011
 8 MIMOSA 26 sensors, 4 as reference planes in the telescope, 4 in the PLUME ladder (2 per side),
 SPS beam 120 GeV pions, normal incidence but beam located in between two PLUME sensors

- RUN 28510, July 2011
 6 MIMOSA 28 (ULTIMATE) sensors in a telescope (4 as references and 2 as DUT), SPS beam 120 GeV pions, normal incidence
 acquisition = new PXIexpress system where only triggered frames are written

- RUN 28905, July 2011
 6 MIMOSA 28 (ULTIMATE) sensors in a telescope but data taken without beam,
 this run is to estimate the fake hit rate
  acquisition = new PXIexpress system where only triggered frames are written

- RUN 26519, August 2010
 6 MIMOSA 26 sensors in a telescope (4 as references and 2 as DUT), SPS beam 120 GeV pions, normal incidence
 acquisition = old PXI system where all frames are written

- RUN 22704,
 1 MIMOSA 22 AHR sensor in a telescope of 8 strip planes, SPS beam 120 GeV pions, normal incidence
 acquisition = Imager-USB system

- RUN 18400,
 1 MIMOSA 18 sensor exposed to a Rutenium source
 acquisition = TNT board

- RUN 18007,
 1 MIMOSA 18 sensor exposed to a Rutenium source
 acquisition = IMAGER board

 - RUN 226020
   PLUME Beam test @ CERN-SPS on Oct. 2011: 8 Mi26 chips.
   acquisition PXIe

 - RUN 28822 and 28848
   SALAT test beam @ DESY on Feb 2014: 8 Mi28 chips
   acquisition PXIe

 - RUN 18538
   Mi18 Beam test @ CERN-SPS on Sep. 2007: Telescope with 4 planes of Mi18 chips.
   acquisition IMG

 - RUN 35507 and 35604
   FSBB-M0 test beam @ CERN-SPS (Oct. 2014) and DESY (Mar. 2015): Telescope of 6 planes of FSBB-M0 chips.
   Acquisition PIXe

 - RUN 35705 and 35604
   FSBB-M0-bis test beam @ DESY (Jun. 2015) and CERN-SPS (Oct. 2015): Telescope of 6 planes of FSBB-M0-bis chips.
   Acquisition PIXe



*********************************************************************************************************
COMMENTS ON FEATURES
*********************************************************************************************************

- compilation
  * PXIeBoardReader.cxx uses an #include to define which pxi_daq library version is used

- Variables which are hardcoded

  * A number of options in IMGBoardReader.

  * In the method find_tracks() of the class DTracker.cxx, you choose by commenting
    the method to extrapolate the track to the next plane.
    --> not true anymore since v2.8.1

  * In DTrack.cxx, method Analyze, you choose to use a vertex point
    to start your track or not by commenting/uncommenting the proper lines.
    --> not true anymore since v2.8.1

  * DAcq.cxx sets the "vetoOverflow" flag for the PXIeBoardReader objects (line 177)

  * Each DAQ Board type is associated with a unique integer id in DAcq.cxx
    1 = IMG, 2 = ?, 3 = TNT, 4 = PXI, 5 = PXIe, 6 = GIG

  * Max number of modules to be associated to a plane is 16 (from release 2.0, 4 in 1.0). This is hardcoded in DSetup.h and can be changed if needed.

  * max number of modules for a given module type is 16
    in DSetup.h for the declaration of *DeviceDataFile[16]

  * Max number of submatrix (in total, whatever the number of DUT) is 10, hardcoded in DSetup.h
    First subamtrix number is 0.

  * limits on shift and tilt change to stop alignment in MAlign.cxx

  * maximum size of the IndexOfGoodHitsInEvent array in MCommands.cxx

  * minimum noise to consider the value is reasonable and hence compute a S/N for a pixel
    min value is 0.5 in DHit::Analyse and in DPixel::GetPulseHeightToNoise

*********************************************************************************************************
TO DO LIST FOR DEVELOPERS
*********************************************************************************************************

* HIGH PRIORITY

- Propose a display of pixel position (based on (col,lin)<->(u,v) methods)

- Allows Mrax class to work on any geometry

- Manage properly and uniformely the MimosaType variable everywhere:
   set it once for all

- Set a method to sparsify the data for analogue output with no noise recomputation

- Reshape MRaw to have histos booked and filled separately from the display
  => might be useless since the introduction of the MRax class

- Reshape the classes DHit and DTrack to allow easy switching between different
   clustering and tracking algorithm.
  => Currently managed with flags in the config file to switch to different find_xxx methods (JB July 2013)

- Insure all variables are stable against 64bits CPU (especially in BoardReader classes)
  => Currently work with a patch...(JB+LC, May 2012)

- Provide a Digitizer that simulates the pixel response to a list of energy
  depositions read from an input file

- Add some statistics computation on the noise distribution

* LOW PRIORITY

- Check in MimosaPro that hits are not selected several times by different tracks

- In each output file, store a canvas with text about the configuration and cuts

- Add protection if DTDIR not defined

- Provide an estimation of the telescope resolution from geometry and available plane resolution

- Management of cuts (both for DSFProduction and for final analyses) especially for bookkeeping

- Create a GetCuts method in MCommands managing one or two planes

- Better management of readout types and mimosa types in DPLane::Update (it is a mess currently)

- Take care of sensor aspect ratio in MRaw.cxx method for display

- In the align tracking procedure, ask the user to store or not the new parameters,
   could be usefull in the case where the fit procedure went wrong

- Improve the MimosaFakerate command to get rate for each submatrix and per column and line

- change the computation of the geoMatrix limit
	from position in microns
	to index (row and column) of pixels

- for eta algorithm in tracker switch from histograms to arrays (like done in MimosaAnalysis)

- include AlignStatus of tracker in config file (maybe useless?)

* DONE

- Allow different event building mode for the DUT and the Ref planes in PXIeBoardReader,
   for instance, DUT event may use one additional frame compared to REF event
  => done, JB 2014

- Introduce a digitization emulation algorithm in DPlane
  => done, JB August 2013

- Add fit of residual (DV/DU) vs (V/U) in Minuit Alignment procedure
  => done, JB August 2013

- Change the reading method for the configuration file in DSetup
  to allow for new variables BUT keeping the compatibility with old files
  => done, JB January 2013

- include the precision alignment of DUT in the method AlignTracker
  => done, LC September 2012

- Cope with strip-telescope data
  => done, JB August 2012

- Provide a ImagerBoardReader to decode rawdata from the IPHC Imager-USB board
  => done, JB+SS August 2012

- Generate a printout method for MCommand.cxx (#splitline{top}{bottom})
  => mostly done, SS+MG 2012/?

- CrossCheck coherence between cuts ClusterLimitU and MaxNStrips
  => done, JB 2012/05/09

- Allow usage of run number with any digits (especially for directory naming)
  => done, JB+VR 2012/05/04

- Provide a MCBoardReader which would read a GEANT4 output file, digitize the energy deposition and then provide a list of pixels as any other BoardReader
  => done JB+LC, 2012/04/24 (GIGBoardReader)

- Allow display of position in tracker coordinates and not only in plane coord. for MRaw Display functions
  => done JB+VR, 2012/04/02

- Move the Hot Hit Map mechanism used in MimosaPro method as a generic methods usable by others MCommands.cxx methods.
  => done JB, 2011/11/23

- Provide a method to deal with clusters containing several occurenres of the same pixel. This happens when several frames are merged to produce a single physical event.
  => done JB+SS, 2011/11/07

- Modify DSFProduction to choose the level of information actually stored (hits of all plane or not, tracks...)
  => done JB 2011/07/21

- Get rid of all global/external variables for compilation under WINDOWS
  => done JB 2011/07/21

- Manage efficiency computation in MCommands.cxx methods as separate and general methods in the MAnalysis.cxx class. Also include efficiency computation dependant on X and Y.
  => done MG+JB 2011/11/04

- Make the generation of directories compatible with any system (add Windows!)
  => done JB 2011/07/18

- Use the Root Documentation generation for the classes.
  => done JB 2011/07/20

- Allow to increment above 9 files for the TNT input files
  => done JB 2010/10/06

- Add new word case in TNT word=50000000 for overflow
  => done JB 2010/10/06

- Include statistics printout for TNTBoardReader
  => done JB 2010/09/27

- Re-shuffle MimosaPro function for clarity, all histogram filling should be externalized
  => done JB 2010/07/29

- mechanism to compute noise from TNT + to retrieve it during data mining
  => done YV 2009/?

- implement histograms for statistics on data mining during the DSFProduction
  => done JB 2009/09/10

- have a mechanism to change the debug level after InitSession
  => done JB 2009/05/12

- read the rawdata filename from config file
  => done JB 2009/05/25

- setup a mechanism to reject event when too many hits in a plane
  => done  JB 2009/08/26

- allow best track over all possible ones when several hits
  => done JB 2009/05/21

- upgrade to new TNT format of May 2009
  => done RDM 2009/05/20

- include harcoded parameters in config file (# events for alignment, max hit-track distance...)
  => done JB 2009/05/25

- Use the globalTool function info, error instead of simple cout or printf
  => probably useless

- Check that all modules declared are actually associated to a plane, otherwise STOP.
  => done JB 2009/05/25

- If possible get rid of the MPara class, everything should come from the config file
  => done JB 2009/08/26


releases.txt
*********************************************************

LIST OF RELEASES (or tagged versions) OF TAF

*********************************************************
Contact: baudot@in2p3.fr, auguste.besson@iphc.cnrs.fr
Web page: http://www.iphc.cnrs.fr/TAF.html

*********************************************************************************************************
Master - 2020/12/03
*********************************************************************************************************

- additional features in Script/thistaf to allow conditional compilation <== does not work fully yet
- documentation updates (Create your own BoardReader, )
- Reader IMGBoardReader updated to read data format for Monolithic Imager sensor (MimosaType 61)

*********************************************************************************************************
VERSION 3.00.0 - 2020/08/20
*********************************************************************************************************

- additional features in MRaw to handle SITRINEO (Student test bench) analysis
- Reader for new VMEBoardReader to adapt for change in INFN readou board

*********************************************************************************************************
Master - 2018/10/17
*********************************************************************************************************

- New reader (BoarReaderIHEP) to read data from IHEP acquisition system
- Possibility to declare dummy plane, with no data associated, just for multiple scattering
- New methods in MRaw for X-rays and image production
- New method Mraw::StudyBeamProfile for beam profile application (thanks Q.Liu)
- Some fixes to comply with clang compilation rules (thanks A.Maire)
- Improvement of deformation correction (included in DPrecAlign now)
- New command in MRaw to study multiple scattering
- New macro  SpatialResolution_with_EnergyScan
- New internal toyMC package to generate tracks (gTAF->MimosaGeneration)
- Method to remove hot pixels more robust (shall work with all type of BoardReader)


*********************************************************************************************************
VERSION 2.10.0 - 2015/01/07
*********************************************************************************************************


- Possibility to get the fit from a subset of hits within a track (SubTrack)
- Track fit now exploits fully the plane resolution,
  which can be set in the configuration file
- Propagate flag to TTree to indicate if hit used in track fit
- A number of corrections in classes to decode E.Spiriti's DAQ output
   (DecoderM18, AliMIMOSA22RawStreamVASingle) from INFN colleagues
- New methods for clustering related to sparse-binary data, by A.Perez
- Improved cluster superposition to pixel raw data in plots from MRaw

*****************************
 New features, by Valérian: 2014/07/01
*****************************
-  PXI DAQ lib configuration centralized in code/include/pxi_daq_lib_config.h
-  New TAF configuration script : thistaf.sh should be used instead of TAF-config
-  New Makefile.arch : see Makefile for help or configuration
-  Makefile displays clearer messages, and get the svn revision for the TAF Welcome message
-  TAF can be lauched with arguments for various actions (like $TAF -run 26005), see help with $TAF --help
-  MRax class, inspired from MRaw is added
-  Tracking method configurable via config file
-  The RunNumber in config file is obsolet, the value from InitSession is now used
-  A configFile can be set via an argument when launching TAF
-  TAF can be launched and work at any place
-  DataSubDirPrefix parameter added to config file reader for data file in subdirs like DataPath/RUN26001/
-  Some Mandatory fields are tested inconfig file, if not present, print an error and quit
- rootlogon.C the explicit loading of TAF & Minuit libraries is configurable
*****************************


- New method for the history of some hits & tracks properties wrt event#
  in MRaw.cxx (gTAF->GetRaw()->DisplayHistory())
- rootlogon.C updated with explicit loading of TAF & Minuit libraries
- New method to generate image with MRaw.cxx (gTAF->GetRaw()->DisplayImage())
- New classes to decode more formats related to E.Spiriti DAQ
- Possibility to parametrize plane deformation
   through Legendre polynomials (see DPlane.cxx)
   with new parameters in config files (see DSetup.cxx)
- Possibility to customize ladder (instead of fixed types)
- Individual pixel distribution available with gTAF->GetRaw()->DisplayNoise()
- Possibility to specify an analysis goal in the configuration file,
   so that not all plots are generated (for MimosaXxxxx command type)
- Bug corrected in find_tracks (hits on seed planes not associated to a track are set un-associated)

*********************************************************************************************************
VERSION 2.9.7 - 2013/11/12
*********************************************************************************************************

- One variable (HsN) for seed SNR replaced in the TTRee (DAuthenticHit)
- New plots for cluster properties, reshuffled in separate method (CheckCluster)
- New methods for alignement with Minivectors from Loic,
  class DPrecAlign has new parameters (CorPar file format is hence changed)
- Possibility to set the DSF file name to any name, not just a number
- Bug fix in file list management

*********************************************************************************************************
VERSION 2.9.6 - 2013/09/23
*********************************************************************************************************

- Method AlignTrackerMinuit includes iterations managed manually
- Range of histograms for noise, signal and SNR set in configuration file
- All histos are written in a single file (runXXXXX_ClChargePlY.root) with MimosaPro
- Digitization emulation in DPlane, controlled with "Digitizer" field in configuration


include/DHit.h
// @(#)maf/dtools:$Name:  $:$Id: DHit.h,v.1 2005/10/02 18:03:46 sha Exp $
// Author   :  ?

#ifndef _DHit_included_
#define _DHit_included_

  ////////////////////////////////////////////////////////////
  //                                                        //
  // Class Description of DHit                              //
  //                                                        //
  ////////////////////////////////////////////////////////////

#include "Riostream.h"
#include <math.h>
#include <vector>

// ROOT classes
#include "TObject.h"
#include "TArrayF.h"
#include "TAxis.h"
#include "DPixel.h"
//#include "DPlane.h"

class DTrack;                  // forward declarations
class DStrip;
class DPlane;
class DCut;
class DR3;

class DHit : public TObject {

 private:
  Int_t              fTableSize;                // Size of Arrays allocated in constructor // VR 2014.08.28
  Int_t              fDebugHit;
  Int_t              fHitNumber;                // hit number
  Bool_t             fFound;                    // kTRUE is associated to a track, JB 2009/05/22
  Int_t              fPositionAlgorithm;        // 1 = Center of Gravity, 2 = eta, 3= kappa, 4 = gamma
  DPlane            *fPlane;                    // pointer to the device which got this hit
  DR3               *fPositionHit;              // position of the hit, depends on fPositionAlgorithm
  DR3               *fPositionHitCG;            // position of the hit measured with center of gravity method
  DR3               *fPositionHitEta;           // position of the hit measured with eta correction from 3x3 cluster
  DR3               *fPositionHitCG33;          // position of the hit measured with center of gravity method over a 3x3 cluster
  DR3               *fPositionHitCG22;          // position of the hit measured with center of gravity method over a 2x2 cluster
  DR3               *fPositionHitEta22;         // position of the hit measured with eta correction from 2x2 cluster
  DR3               *fResolutionHit;            // position resolution of the hit, depends on resolution estimation, AP 24/11/2014
  Float_t            fSeedU;                    // U position of hit seed strip
  Float_t            fSeedV;                    // V position of hit seed strip
  Float_t            fClusterPulseSum;          // sum of pulseheight on strips in hit cluster, involves noise cuts
  Float_t            fClusterAreaPulseSum;      // sum of pulseheight on strips in hit cluster area, no noise cuts
  Float_t            fClusterSignalToNoise;     // hit cluster signal to noise ratio
  Float_t            fClusterNoiseAverage;      // hit cluster signal noise average
  Int_t              fStripsInClusterFound;     // number of strips in the hit cluster
  Int_t              fStripsInClusterArea;       // # strips in cluster area
  DR3               *fClusterLimit;             // maximum extension in u,v,w for hit analysis
  Float_t           fClusterLimitRadius;        // maximum search radius (in mm) from real center of gravity (in mm) to associate new pixels
  DStrip            *fSeed;                     // pointer to the hit seed strip
  DPixel            *fPSeed;                    // pointer to the hit seed pixel, JB 2009/05/01
  //DMonteCarlo       *fPSeedMC;                  // pointer to the pixel Monte Carlo, LC 2014/12/15.
  DCut              *fCut;                      // pointer to cuts
  DR3               *fStripPitch;               // pointer to strip pitch in plane
  Float_t            fPhSeed;                   // pulseheight on seed strip
  Float_t            fPosSeed;                  // position of the seed strip
  Float_t            fPhLofSeed;                // pulseheight on strip left from seed
  Float_t            fPosLofSeed;               // position of strip left of seed
  Float_t            fPhRofSeed;                // pulseheight on strip right from seed
  Float_t            fPosRofSeed;               // position of strip right from seed
  Int_t              fIndexSeed;                // index of seed strip
  Int_t              fIndexLofSeed;             // index of strip left from seed
  Int_t              fIndexRofSeed;             // index of strip right from seed
  Float_t            fPhLeft;                   // pulseheight on left most strip in two strip cluster
  Float_t            fPosLeft;                  // position of left most strip in two strip cluster
  Float_t            fPhRight;                  // pulseheight on right most strip in two strip cluster
  Float_t            fPosRight;                 // position of strip in right most cluster
  Int_t              fIndexLeft;                // index of left most strip
  Int_t              fIndexRight;               // index of right most strip
  Int_t              fStoNover2;                // MB 2010/11/12

  Float_t           *fStripNoise;               //! array of noise from strips in hit cluster  YV 27/11/09
  Int_t             *tPixelIndexList;           //! array of pixel/strip index in the original list of pixels
  Int_t             *fStripIndexArray;          //! array of indices to strips inside the hit
  Int_t             *fStripIndex;               //| array of index of strip in the full plane matrix
  Float_t           *fStripPulseHeight;         //! array of pulseheights from strips in hit cluster
  Float_t           *fStripDistanceU;           //! array of seed distanceU from strips in hit cluster
  Float_t           *fStripDistanceV;           //! array of seed distanceV from strips in hit cluster
  Int_t              fStripsInClusterDemanded;  // exact number of strips required in Hit, set by user
  Float_t            evaluateChargeFraction(Float_t aChargeFraction);  // evaluates the charge fraction on strip
  TArrayF           *fChargeFractionDensity;    // Array of charge fraction densities for hit measurement

  Float_t           fSNneighbour;               // signal / noise of neighbours
  Float_t           fSNseed;                    // signal / noise of seed pixel, JB 2013/11/08
  Float_t           fSeedPulseHeight;           // pulseheight on seed, JB 2013/11/08
  Float_t           fSeedNoise;                 // noise on seed, JB 2013/11/08
  Bool_t            fIsFromPreviousEvent;       // explicit // VR 2014.08.28
  Int_t             fTimestamp;                 // timestamp of seed, JB 2015/05/25
  Bool_t            fIfMonteCarlo;
  std::vector<Double_t> _monteCarloInfo;         // Vector of Monte Carlo Informations. Vector to reduce memory if no MC Info + Add new params easily
  Int_t             fMCHitID;                   // MC hit ID number. This number gives MC hit ID generating the hit. Default value is -1
  Int_t             fStripsFromMCHitID;         // This gives the number of strips/pixels from the MC hit with ID = fMCHitID. Default value if 0

  //Main components analysis variables
  Float_t           fNormMaxStd;                //RMS of pixel position along principal axis
  Float_t           fNormMinStd;                //RMS of pixel position along axis perpendicualer to principal axis
  Float_t           fMainAxisPhi;               //Phi angle of principal axis

  int               fColMin;                    //Pixel with the lowest  column value
  int               fColMax;                    //Pixel with the highest column value
  int               fRowMin;                    //Pixel with the lowest  row    value
  int               fRowMax;                    //Pixel with the highest row    value

public:
  DHit();
  DHit(DR3 &aPosition, DPlane& aPlane, Int_t aHitNumber);
  DHit(DR3 &aPosition, DPlane& aPlane, Int_t aHitNumber, std::vector<Double_t>& monteCarloVector);
  void               copy(DHit* aHit, DPlane* aPlane); // LC 2014/07/15 // Not full a copy :: just hit positions :)
  void               clone(DHit *original, Bool_t fullCopy=1);// VR 2014.08.28, cannot call it Clone otherwise hidden virtual function
  //DHit(DR3 &aPosition, Int_t aHitNumber);
  virtual           ~DHit();
  Bool_t             Analyse(DStrip *s);            // cluster charge, noise, hit position
  Bool_t             Analyse_Iterative(DStrip *s, bool &IsBigCluster, int MaxClusterSize);  //AP, 07/01/2014, Iterative clustering for digital output
  Bool_t             Analyse( Int_t aPixelIndexInList, std::vector<DPixel*> *aListOfPixels);   // cluster charge, noise, hit position
  Bool_t             Analyse_2_cgo( Int_t aPixelIndexInList, std::vector<DPixel*> *aListOfPixels); // make hit with pixels considering maximum search radius (in mm) from real center of gravity (in mm) to associate new pixels
  Bool_t             Analyse_Iterative( Int_t aPixelIndexInList, std::vector<DPixel*> *aListOfPixels, bool &IsBigCluster, int MaxClusterSize);   //AP, 07/01/2014, Iterative clustering for digital output
  //Bool_t             AnalyseMC( std::vector<DMonteCarlo*> *aListOfPixelsMonteCarlo)                    //LC, 2014/12/15, Get MC Hit position
  Int_t              GetNumber()                      const { return  fHitNumber;            }
  void               SetNumber(Int_t aNb)                   { fHitNumber = aNb;}// VR 2014.08.28
  Bool_t             GetFound()                       const { return fFound; }
  void               SetFound( Bool_t found=kTRUE)          { fFound = found; } // JB 2011/03/14 to allow to free the hit
  Int_t              GetDebug()                             { return fDebugHit;              }
  void               SetDebug( Int_t aDebug)                { fDebugHit = aDebug;            }
  DPixel            *GetPSeed()                             { return  fPSeed;                } // JB 2012/08/17
  DStrip            *GetSeed()                              { return  fSeed;                 }
  DStrip            *GetMinor(Int_t aSk);             // strip with index lower pulseheight in neighourhood to seed
  DR3               *GetPosition()                    const { return fPositionHit;}
  DR3               *GetPositionCG()                  const { return fPositionHitCG;}
  Float_t            GetPositionUhit() const;
  Float_t            GetPositionVhit() const;
  Float_t            GetPositionWhit() const;
  Float_t            GetPositionUhitCG() const;
  Float_t            GetPositionVhitCG() const;
  Float_t            GetPositionWhitCG() const;
  Float_t            GetPositionUhitEta() const;
  Float_t            GetPositionVhitEta() const;
  Float_t            GetPositionUhitCG33() const;
  Float_t            GetPositionVhitCG33() const;
  Float_t            GetPositionUhitCG22() const;
  Float_t            GetPositionVhitCG22() const;
  Float_t            GetPositionUhitEta22() const;
  Float_t            GetPositionVhitEta22() const;
  Float_t            GetResolutionUhit() const;
  Float_t            GetResolutionVhit() const;
  void               SetResolutionUVhit(Float_t resolutionU,Float_t resolutionV);
  Float_t            GetClusterPulseSum()             const { return  fClusterPulseSum;      }
  Float_t            GetClusterAreaPulseSum()         const { return  fClusterAreaPulseSum;  }
  Int_t              GetIndex(Int_t tSk)              const { return  fStripIndex[fStripIndexArray[tSk]];      } // JB 2009/05/12
  Int_t              GetIndexInOriginalList(Int_t tSk) const { return tPixelIndexList[tSk]; }  //AP 2016/07/28
  Bool_t             GetIsFromPreviousEvent()         const { return fIsFromPreviousEvent;}// VR 2014.08.28
  void               SetIsFromPreviousEvent(Bool_t value=1) {fIsFromPreviousEvent=value;}// VR 2014.08.28
  Float_t            GetPulseHeight(Int_t tSk)        const { return  fStripPulseHeight[fStripIndexArray[tSk]];} // JB 2009/05/12
  Float_t            GetNoise(Int_t tSk)              const { return  fStripNoise[fStripIndexArray[tSk]];}  //YV 27/11/09
  Float_t            GetClusterSignalToNoise()        const { return  fClusterSignalToNoise; }
  Float_t            GetClusterNoiseAverage()         const { return  fClusterNoiseAverage;  }
  Int_t              GetStripsInCluster()             const { return  fStripsInClusterFound; }
  DPlane            *GetPlane()                       const { return  fPlane;                }
  Float_t            GetSeedU()                       const { return  fSeedU;                }
  Float_t            GetSeedV()                       const { return  fSeedV;                }
  Float_t            GetPulseHeightLeft()             const { return  fPhLeft;               }
  Float_t            GetPulseHeightRight()            const { return  fPhRight;              }
  Float_t            GetPulseHeightLeftOfSeed()       const { return  fPhLofSeed;            }
  Float_t            GetPulseHeightRightOfSeed()      const { return  fPhRofSeed;            }
  Float_t            GetPositionULeft()               const { return  fPosLeft;              }
  Int_t              GetStoNover2()                   const { return  fStoNover2;            }

  Int_t              GetIndexLeft()                   const { return  fIndexLeft;            }
  Int_t              GetIndexRight()                  const { return  fIndexRight;           }
  Int_t              GetIndexSeed()                   const { return  fIndexSeed;            }

  Float_t            GetSNneighbour()                 const { return  fSNneighbour;         }
  Float_t            GetSNseed()                      const { return  fSNseed;          } // JB 2013/11/08
  Float_t            GetSeedPulseHeight()             const { return  fSeedPulseHeight;          } // JB 2013/11/08
  Float_t            GetSeedNoise()                   const { return  fSeedNoise;          } // JB 2013/11/08

  Float_t            Distance( DHit *aHit);
  Float_t            Distance( DTrack *aTrack); // JB 2009/06/26
  Float_t            Distance( DR3 *aPoint); // JB 2011/07/25

  Int_t              GetTimestamp()                   const { return fTimestamp; }; // JB 2015/05/25
  
  Bool_t             IfMonteCarlo()                   const { return fIfMonteCarlo; }; // LC 2014/12/15 -> Test if MC Infos are included.

  Float_t            DistanceMC( DHit *aHit);     // LC 2014/12/15
  Float_t            DistanceMC( DTrack *aTrack); // LC 2014/12/15
  Float_t            DistanceMC( DR3 *aPoint);    // LC 2014/12/15

  void               SetMCHitID(Int_t aHitID) {fMCHitID = aHitID;}  //AP 2015/07/30
  Int_t              GetMCHitID()             {return fMCHitID;}    //AP 2015/07/30

  void               SetStripsFromMCHitID(Int_t aStripsFromMCHitID) {fStripsFromMCHitID = aStripsFromMCHitID;}  //AP 2016/07/28
  Int_t              GetStripsFromMCHitID()                         {return fStripsFromMCHitID;}                //AP 2016/07/28

  void               SetHitPosition(DR3 aPosition,DR3 aResolution);

  DR3                GetMonteCarloPosition();
  std::vector<Double_t>& GetMonteCarloInfo()          { return _monteCarloInfo; } // LC 2014/12/15

  //Main components analysis functions
  void               DoMCA(void);  //AP 2017/05/09
  Float_t            GetNormMinStd()                  { return  fNormMinStd;  }  //AP 2017/05/09
  Float_t            GetNormMaxStd()                  { return  fNormMaxStd;  }  //AP 2017/05/09
  Float_t            GetMainAxisPhi()                 { return  fMainAxisPhi; }  //AP 2017/05/09

  void               CalcHitLimits(void); //AP 2017/05/11
  int                GetColMin()                      { return fColMin; }  //AP 2017/05/11
  int                GetColMax()                      { return fColMax; }  //AP 2017/05/11
  int                GetRowMin()                      { return fRowMin; }  //AP 2017/05/11
  int                GetRowMax()                      { return fRowMax; }  //AP 2017/05/11

  Bool_t             IsSortable() const { return kTRUE; } // QL 04/06/2016
  Int_t              Compare( const TObject * obj) const; // QL 04/06/2016
  virtual void       Print(const Option_t* ="") const; // QL 05/06/2016

  ClassDef(DHit,1)                              // Hit in a Plane
};

#endif
